libraries/lib-transactions/TransactionScope.cpp

```diff
@@ -0,0 +1,101 @@
+/*!********************************************************************
+
+  Audacity: A Digital Audio Editor
+
+  @file TransactionScope.cpp
+
+  Paul Licameli split from DBConnection.cpp
+
+**********************************************************************/
+
+#include "TransactionScope.h"
+#include "InconsistencyException.h"
+#include <wx/log.h>
+
+namespace {
+TransactionScope::TransactionScopeImplFactory &GetFactory()
+{
+   static TransactionScope::TransactionScopeImplFactory theFactory;
+   return theFactory;
+}
+
+TransactionScope::AutoSaveFunction &GetAutoSave()
+{
+   static TransactionScope::AutoSaveFunction theFunction;
+   return theFunction;
+}
+}
+
+auto TransactionScope::InstallImplementation(
+   TransactionScopeImplFactory factory) -> TransactionScopeImplFactory
+{
+   auto &theFactory = GetFactory();
+   auto result = theFactory;
+   theFactory = move(factory);
+   return result;
+}
+
+auto TransactionScope::InstallAutoSave(
+   AutoSaveFunction function) -> AutoSaveFunction
+{
+   auto &theFunction = GetAutoSave();
+   auto result = theFunction;
+   theFunction = move(function);
+   return theFunction;
+}
+
+void TransactionScope::AutoSave(AudacityProject &project)
+{
+   auto &function = GetAutoSave();
+   if (function)
+      function(project);
+}
+
+TransactionScopeImpl::~TransactionScopeImpl() = default;
+
+TransactionScope::TransactionScope(
+   AudacityProject &project, const char *name)
+:  mName(name)
+{
+   auto &factory = GetFactory();
+   if (!factory)
+      return;
+   mpImpl = factory(project);
+   if (!mpImpl)
+      return;
+
+   mInTrans = mpImpl->TransactionStart(mName);
+   if ( !mInTrans )
+      // To do, improve the message
+      throw SimpleMessageBoxException( ExceptionType::Internal,
+         XO("Database error.  Sorry, but we don't have more details."), 
+         XO("Warning"), 
+         "Error:_Disk_full_or_not_writable"
+      );
+}
+
+TransactionScope::~TransactionScope()
+{
+   if (mpImpl && mInTrans)
+   {
+      if (!mpImpl->TransactionRollback(mName))
+      {
+         // Do not throw from a destructor!
```

Comment:
This feels like a possible source of broken projects, although rollback failures are rather rare

Reply:
Rollback as the RAII step in a desctructor makes sense.  A destructor should also not throw.  It should do no-fail operations.  Yet we can't quite be sure rollback is always no-fail?  (A decent database should try very hard to make it so!)

No totally acceptable thing to do for this rare eventuality.  But again, it's beside the purpose of this PR, which does not make this problem better or worse than it was.

Reply:
>A destructor should also not throw

This is true 99.9% of the time, but probably this one is in 0.1%? :-) C++ really fails us here.

I think we should rewrite transactions in the future (?) to accept the function object, that should be run as a part of the transaction. Like `io->Transaction([this](UndoStuff();});`